src/server/config_registry.cc

```diff
@@ -87,6 +97,136 @@ void ConfigRegistry::RegisterInternal(string_view config_name, bool is_mutable,
   CHECK(inserted) << "Duplicate config name: " << name;
 }
 
+bool ConfigRegistry::Rewrite() const {
```

Comment:
lets separate the config file generation from writing it to the file. Right now it's all mixed together and it's hard to follow.
Introduce another function `std::string GenerateConfig() const` that only generates the string contents you want to serialize. And then  write these contents to the file.